### PR TITLE
[SPARK-25388][Test][SQL] Detect incorrect nullable of DataType in the result

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR can correctly cause assertion failure when incorrect nullable of DataType in the result is generated by a target function to be tested.

Let us think the following example. In the future, a developer would write incorrect code that returns unexpected result. We have to correctly cause fail in this test since `valueContainsNull=false` while `expr` includes `null`. However, without this PR, this test passes. This PR can correctly cause fail.

```
test("test TARGETFUNCTON") {
  val expr = TARGETMAPFUNCTON()
  // expr = UnsafeMap(3 -> 6, 7 -> null)
  // expr.dataType = (IntegerType, IntegerType, false)

  expected = Map(3 -> 6, 7 -> null)
  checkEvaluation(expr, expected)
```

In [`checkEvaluationWithUnsafeProjection`](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L208-L235), the results are compared using `UnsafeRow`. When the given `expected` is [converted](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L226-L227)) to `UnsafeRow` using the `DataType` of `expr`.
```
val expectedRow = UnsafeProjection.create(Array(expression.dataType, expression.dataType)).apply(lit)
```

In summary, `expr` is `[0,1800000038,5000000038,18,2,0,700000003,2,0,6,18,2,0,700000003,2,0,6]` with and w/o this PR. `expected` is converted to

* w/o  this PR, `[0,1800000038,5000000038,18,2,0,700000003,2,0,6,18,2,0,700000003,2,0,6]`
* with this PR, `[0,1800000038,5000000038,18,2,0,700000003,2,2,6,18,2,0,700000003,2,2,6]`

As a result, w/o this PR, the test unexpectedly passes.

This is because, w/o this PR, based on given `dataType`, generated code of projection for `expected` avoids to set nullbit.
```
                    // tmpInput_2 is expected
/* 155 */           for (int index_1 = 0; index_1 < numElements_1; index_1++) {
/* 156 */             mutableStateArray_1[1].write(index_1, tmpInput_2.getInt(index_1));
/* 157 */           }
```

With this PR, generated code of projection for `expected` always checks whether nullbit should be set by `isNullAt`
```
                    // tmpInput_2 is expected
/* 161 */           for (int index_1 = 0; index_1 < numElements_1; index_1++) {
/* 162 */
/* 163 */             if (tmpInput_2.isNullAt(index_1)) {
/* 164 */               mutableStateArray_1[1].setNull4Bytes(index_1);
/* 165 */             } else {
/* 166 */               mutableStateArray_1[1].write(index_1, tmpInput_2.getInt(index_1));
/* 167 */             }
/* 168 */
/* 169 */           }
```

## How was this patch tested?

Existing UTs

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 1]
```diff
@@ -223,9 +223,9 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
           }
         } else {
           val lit = InternalRow(expected, expected)
-          val expectedRow =
-            UnsafeProjection.create(Array(expression.dataType, expression.dataType)).apply(lit)
-          if (unsafeRow != expectedRow) {
+          val dtAsNullable = expression.dataType.asNullable
+          val expectedRow = UnsafeProjection.create(Array(dtAsNullable, dtAsNullable)).apply(lit)
+          if (unsafeRow == expectedRow) {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Why did you remove the existing test `unsafeRow != expectedRow`?

Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Thank you, it was used for debugging.

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: ah, ok.

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 2]
```diff
@@ -223,8 +223,8 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
           }
         } else {
           val lit = InternalRow(expected, expected)
-          val expectedRow =
-            UnsafeProjection.create(Array(expression.dataType, expression.dataType)).apply(lit)
+          val dtAsNullable = expression.dataType.asNullable
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Add some comments explaining it? Not so straightforward.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 3]
```diff
@@ -223,8 +223,8 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
...
+          val expectedRow = UnsafeProjection.create(Array(dtAsNullable, dtAsNullable)).apply(lit)
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Can we add a test for this change?

Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Thank you for your comment. Do you think what test is preferable?
* Successfully pass (we already have these UTs since all UTs have been passed)
* Expectedly failure (In other words, add a function that generated incorrect result)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 2 above)
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Sure, add some comments from the description

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: I'd like to add a test that will be failed without this patch.

Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Since this patch allows us to correctly detect incorrect results, I added a test that was passed w/o this patch and that is failed with this patch.

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: oh, yes. this is going to detect such failure.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala [hunk 4]
```diff
@@ -35,6 +36,13 @@ class ExpressionEvalHelperSuite extends SparkFunSuite with ExpressionEvalHelper
     val e = intercept[RuntimeException] { checkEvaluation(BadCodegenExpression(), 10) }
     assert(e.getMessage.contains("some_variable"))
   }
+
+  test("SPARK-25388: checkEvaluation should fail if nullable in DataType is incorrect") {
+    val e = intercept[RuntimeException] {
+      checkEvaluation(MapIncorrectDataTypeExpression(), Map(3 -> 7, 6 -> null))
```
[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Here is an output. This is because the test correctly detects a failure even in codegen-off mode, too.

```
"Incorrect evaluation (codegen off): mapincorrectdatatypeexpression(), actual: keys: [3,6], values: [7,null], expected: keys: [3,6], values: [7,-1]" did not contain "Incorrect evaluation in unsafe mode"
ScalaTestFailureLocation: org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelperSuite$$anonfun$4 at (ExpressionEvalHelperSuite.scala:44)
org.scalatest.exceptions.TestFailedException: "Incorrect evaluation (codegen off): mapincorrectdatatypeexpression(), actual: keys: [3,6], values: [7,null], expected: keys: [3,6], values: [7,-1]" did not contain "Incorrect evaluation in unsafe mode"
	at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:528)
...
```

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: I see, so the example above passes in codegen off and fails with codegen on with this fix, while using `Map(3 -> 7, 6 -> -1)` passes codegen on and fails codegen off, am I right?

What I am thinking about (but I have not yet found a working implementation) is: since the problem arise when we say we expect `null` in a non-nullable datatype, can we add such a check? I mean, instead of pretending the expected value to be nullable, can't we add a check in case it is not nullable for being sure that it does not contain `null`? I think it would be better, because we would be able to distinguish a failure caused by a bad test, ie. a test written wrongly, from a UT failure caused by a bug in what we are testing. What do you think?

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Here is summary.
* `Map(3 -> 7, 6 -> null)` passes in codegen off and fails in codegen on with this fix
* `Map(3 -> 7, 6 -> -1)` fails in codegen off and fails in codegen on 

Would it be possible to share examples of two cases that you think we would be able to distinguish?

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: yes, thanks for the summary, it states more clearly what I thought.

My point is that this fix works properly only when we test both codegen on and off, but it would fail to detect the error condition it claims to fix if only one of them (for any reason) is tested. So I am wondering if it is possible to perform a check on the expected value, instead of this fix. Something like:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
where `containsNull` and `isNullable` have to be defined properly. In this way we should fail properly independently from whether codegen is on or not. And we can also give a more clear hint in the error message about the problem being most likely a bad UT.

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Would it be possible to let us know how can this assertion distinguish the following two cases?
* a failure caused by a bad test, ie. a test written wrongly
* a UT failure caused by a bug in what we are testing

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Well, actually I was not clear about this, sorry for that. I meant we can separate from:
 - a failure caused by a wrong result returned;
 - a failure caused by a wrong result type returned (ie. the expected values contains NULL while the expression return data type is not nullable - this case can happen both in case the UT is bad and in case the return type of the expression is wrong);

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: I meant how your proposed assert can distinguish the two cases that you want to distinguish at [here](https://github.com/apache/spark/pull/22375#discussion_r217495874).
1. a failure caused by a bad test, ie. a test written wrongly
2. a UT failure caused by a bug in what we are testing

As you said [here](https://github.com/apache/spark/pull/22375#discussion_r218362156), the assert may detect both cases 1. and 2..

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Sorry, you're right, I have not been precise in my previous statements so they are confusing. I'll try to elaborate more clearly.

IIUC what you are proposing now is for detecting the case when the output datatype of an expression is not nullable and the expected output instead contains a null. The way you're trying to achieve this is to make always nullable the expected output, so that we get a null as output both for codegen on and codegen off (while before this change for codegen on we would get as expected output a wrong value, which is the default value for that data type).

The problems I see on the approach above are:
 - if you test only codegen on or codegen off, you might still be missing to cover some cases (eg. see https://github.com/apache/spark/pull/22375#discussion_r218085246, with codegen off we can have null as output for a non-nullable expression...);
 - The failure you get is that the expected output is different from the the evaluated one, so this case is treated as any other failure in the expression evaluation.

With something like this:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
I think we could solve the 2 problems above as:
 - this check would fail in both modes;
 - we can differentiate between a test failing because of a wrong output and this new case which covers: a bad written UT; an expression returning wrongly a non-nullable datatype instead of a nullable one when returning null.

Hope now it is more clear. Thanks.

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: The your first is correct since this patch addresses only codegen-on case. We can add another code to address codegen-off case.
Regarding the your second point, have we ever distingished a wrong output from a bad written UT when we defect the difference between `expression` and `expected`. I think that the distinguishment is nice to have, but not mandatory to have.


I have one question about your approach:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
Since the above two conditions evaluates `expected` and `expression` independently, how this works for the following case? I think that the assertion would be passed
```
expression:
dataType = StructType(ArrayType(IntegerType, false), ArrayType(IntegerType, true))
Struct(Array(0, null), Array(1, 0))

expected:
Struct(Array(0, 0), Array(1, null))
```

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Yes, you're right, my suggestion doesn't work in a case like that, sorry. We would need to make it checking recursively. But I think you got the idea of what I am proposing here. Thanks.

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: Even if we make it checking recursively, I think that this case cannot be detected. This is because the mismatch occurs in the different recursive path.

Would it be possible to share the case where we distingished a wrong output from a bad written UT in other places, as you proposed?

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: I may not still understand your motivation correctly. What is the motivation to introduce this assertion?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 5]
```diff
@@ -223,8 +223,16 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
           }
         } else {
           val lit = InternalRow(expected, expected)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: I think a more straightforward approach is, validate the `expected` according to the nullability of the given expression.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: The motivations are the 2 mentioned above. Basically, I am proposing the same suggestion @cloud-fan has just commented [here](https://github.com/apache/spark/pull/22375#discussion_r219452615)

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala]: With some hints from @ueshin, this PR implemented the check of `null` value with nullable bit in `checkResult()`.

**Comment by kiszk** [comment]
ping @cloud-fan @mgaido91

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 6]
```diff
@@ -69,11 +69,17 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
...
    */
-  protected def checkResult(result: Any, expected: Any, exprDataType: DataType): Boolean = {
+  protected def checkResult(
+      result: Any,
+      expected: Any,
+      exprDataType: DataType,
+      exprNullable: Boolean): Boolean = {
     val dataType = UserDefinedType.sqlType(exprDataType)
 
+    assert(result != null || exprNullable)
```
Comment (mgaido91) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Can we add a description which is more clear about the issue? Something like: `The result is null for a non-nullable expression`?

Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Sure, I will add the description

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 7]
```diff
@@ -221,6 +227,12 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
         val unsafeRow = evaluateWithUnsafeProjection(expression, inputRow)
         val input = if (inputRow == EmptyRow) "" else s", input: $inputRow"
 
+        val dataType = expression.dataType
+        if (!checkResult(unsafeRow.get(0, dataType), expected, dataType, expression.nullable)) {
```
[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: This is because this statement checks consistency between `expression` and its `nullable`, as you proposed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala [hunk 8]
```diff
@@ -113,7 +113,7 @@ class CodeGenerationSuite extends SparkFunSuite with ExpressionEvalHelper {
     assert(actual.length == 1)
     val expected = UTF8String.fromString("abc")
 
-    if (!checkResult(actual.head, expected, expressions.head.dataType)) {
+    if (!checkResult(actual.head, expected, expressions.head.dataType, expressions.head.nullable)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: It's a little weird to ask the caller to provide both `expected` and `exprNullable `, and then use `exprNullable ` to validate `expected.` Can we set a default value for `exprNullable` in `checkResult`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: mmmh, I am not sure about this. Do we then still need the code below? Seems to me we are checking the same thing twice, please correct me if I am wrong.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala (hunk 8 above)
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: That is another option that I thought. On the other hand, to set default has a risk to overlook a possible incosistency between value and `nullable` at top level of `expected`.

Do we use the default value at the all of callers of `checkResult`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: yes, I just meant that here we are checking the result and we are doing the same after too. Shouldn't we just add an assert for `unsafeRow.get(0, dataType) != null || expression.nullable` here instead?

[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: 1. sees only `expected`. 2. sees `expected` and `expression`. Thus, we are doing different.

At 1, as we discussed, we need to check the consistency recursively. IIUC, `unsafeRow.get(0, dataType) != null || expression.nullable` does not perform checks recursively. Do I make a misunderstanding?

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: > does not perform checks recursively

good point, I was not considering it. Then, do we need the check at https://github.com/apache/spark/pull/22375/files/9ef335d6e43a6ef7d253d0ed3564f95bd0278f71#diff-41747ec3f56901eb7bfb95d2a217e94dL231? Isn't it performed in `checkResult`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: maybe we should provide an overload of `checkResult` that takes `Expression`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (kiszk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: I think `checkResult` already validates `expression` according to the nullability of the given expression at 1. Thus, if the `expected` is not correct, 2. will detect an incorrect point.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 9]
```diff
@@ -69,11 +69,22 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
 
   /**
    * Check the equality between result of expression and expected value, it will handle
-   * Array[Byte], Spread[Double], MapData and Row.
+   * Array[Byte], Spread[Double], MapData and Row. Also check whether exprNullable is true
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: the comment doesn't match the method now

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: we should add message to the assert, e.g. `assert(result != null || exprNullable, "xxx")`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 10]
```diff
@@ -69,11 +69,22 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
...
+
+  protected def checkResult(
+      result: Any,
+      expected: Any,
+      exprDataType: DataType,
+      exprNullable: Boolean): Boolean = {
     val dataType = UserDefinedType.sqlType(exprDataType)
 
+    // The result is null for a non-nullable expression
+    assert(result != null || exprNullable, "exprNullable should be true if result is null")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: nit: how about "result cannot be null since it's not nullable."

**Comment by cloud-fan** [comment]
thanks, merging to master!
